docs/developers/tutorials/staking-contract.md

```diff
@@ -0,0 +1,1416 @@
+
+```rust
+pub fn update<R, F: FnOnce(&mut T) -> R>(&self, f: F) -> R {
+    let mut value = self.get();
+    let result = f(&mut value);
+    self.set(value);
+    result
+}
+```
+
+### Optional arguments
+
+For a bit of performance enhancement, we could have the `unstake_amount` as an optional argument, with the default being full unstake.
+
+```rust
+#[endpoint]
+fn unstake(&self, opt_unstake_amount: OptionalValue<BigUint>) {
+    let caller = self.blockchain().get_caller();
+    let stake_mapper = self.staking_position(&caller);
+    let unstake_amount = match opt_unstake_amount {
+        OptionalValue::Some(amt) => amt,
+        OptionalValue::None => stake_mapper.get(),
+    };
+
+    let remaining_stake = stake_mapper.update(|staked_amount| {
+        require!(
+            unstake_amount > 0 && unstake_amount <= *staked_amount,
+            "Invalid unstake amount"
+        );
+        *staked_amount -= &unstake_amount;
+
+        staked_amount.clone()
+    });
+    if remaining_stake == 0 {
+        self.staked_addresses().swap_remove(&caller);
+    }
+
+    self.send().direct_egld(&caller, &unstake_amount);
+}
+```
+
+This makes it so if someone wants to perform a full unstake, they can simply not give the argument at all.
+
+### Unstaking our devnet tokens
+
+Now that we've added the unstake function, let's test it out on devnet. Build your SC again through the Elrond IDE extension or erdpy directly, and add the unstake function to our snippets.rs file:
+
+```bash
+UNSTAKE_AMOUNT=500000000000000000
+
+unstake() {
+    erdpy --verbose contract call ${SC_ADDRESS} \
+    --proxy=${PROXY} --chain=${CHAIN_ID} \
+    --send --recall-nonce --pem=${USER_PEM} \
+    --gas-limit=10000000 \
+    --function="unstake" \
+    --arguments ${UNSTAKE_AMOUNT}
+}
+```
+
+Now run this function, and you'll get this result:  
+![img](/developers/staking-contract-tutorial-img/first_unstake.png)
+
+...but why? We just added the function! Well, we might've added it to our code, but the contract on the devnet still has our old code. So, how do we upload our new code?
+
+## Upgrading smart contracts
+
+Since we've added some new functionality, we also want to update the currently deployed implementation. Add the upgrade snippet to your snippets.sh and run it:
+
+```bash
+upgrade() {
+    erdpy --verbose contract upgrade ${SC_ADDRESS} \
+    --project=${PROJECT} \
+    --recall-nonce --pem=${USER_PEM} \
+    --gas-limit=20000000 \
+    --send --outfile="upgrade-devnet.interaction.json" \
+    --proxy=${PROXY} --chain=${CHAIN_ID} || return
+}
+```
+
+:::note
+Keep in mind the `#[init]` function of the newly uploaded code is also called on upgrade. For now, it does not matter, as our init function does nothing, but it's worth keeping in mind.
+:::
+
+:::note
+All the storage is kept on upgrade, so make sure any storage changes you make to storage mapping are backwards compatible!
```

Comment:
I think this is an important fact, maybe make it more visible with an "important" or a "warning" layout.